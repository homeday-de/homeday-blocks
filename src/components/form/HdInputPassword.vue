<template>
  <div class="password-input">
    <HdInput
      ref="input"
      :type="type"
      v-model="computedValue"
      v-bind="$attrs"
      :class="{ 'field--hasControl': this.showVisibilityToggle }"
    />

    <span v-if="showVisibilityToggle"
      class="field__visibilityToggle"
      :class="{'field__visibilityToggle--visible': type === 'text'}"
      @click="toggleVisibility"
    >
      <HdIcon
        :src="visibilityOnIcon"
        width="100%"
        height="100%"
      />
    </span>
  </div>
</template>

<script>
import HdInput from 'homeday-blocks/src/components/form/HdInput.vue';
import HdIcon from 'homeday-blocks/src/components/HdIcon.vue';
import { visibilityOnIcon } from 'homeday-blocks/src/assets/small-icons';

export default {
  name: 'HdInputPassword',
  inheritAttrs: false,
  components: {
    HdInput,
    HdIcon,
  },
  props: {
    value: {
      type: String,
      default: '',
    },
  },
  data() {
    return {
      type: 'password',
      visibilityOnIcon,
    };
  },
  computed: {
    computedValue: {
      get() { return this.value; },
      set(value) {
        this.$emit('input', value);
      },
    },
    showVisibilityToggle() {
      return this.value.length !== 0;
    },
  },
  methods: {
    toggleVisibility() {
      this.type = this.type === 'password' ? 'text' : 'password';
    },
    showError(...args) {
      return this.$refs.input.showError(...args);
    },
    showHelper(...args) {
      return this.$refs.input.showHelper(...args);
    },
    validate(...args) {
      return this.$refs.input.validate(...args);
    },
    hideError(...args) {
      return this.$refs.input.hideError(...args);
    },
  },
};
</script>

<!-- Add "scoped" attribute to limit CSS to this component only -->
<style scoped lang="scss">
@import 'homeday-blocks/src/styles/mixins.scss';
@import 'homeday-blocks/src/styles/inputs.scss';

$icon-color: getShade($quaternary-color, 80);

.passwordInput {
  position: relative;

  .field__visibilityToggle {

    position: absolute;
    top: $inputPaddingTop;
    right: $inline-s;
    width: 24px;
    height: 24px;
    cursor: pointer;
    &:after {
      display: block;
      content: '';
      position: absolute;
      opacity: 1;
      top: 2px;
      left: 3px;
      width: calc(100% - 1px);
      height: 0px;
      border-bottom: 2px solid $icon-color;
      border-top: 1px solid $inputBackground;
      transform-origin: left;
      transform: rotateZ(45deg) scaleX(1);
      transition: transform .3s;
    }
    &--visible {
      &:after {
        transform: rotateZ(45deg) scaleX(0);
      }
    }

    &::v-deep path {
      fill: $icon-color;
    }
  }
}
</style>
