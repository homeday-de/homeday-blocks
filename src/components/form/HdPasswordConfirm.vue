<template>
  <div class="confirmPassword">
    <hd-input
      ref="passwordMain"
      v-model="passwordMain"
      name="passwordMain"
      class="confirm-password__input"
      type="password"
      :icon="icon"
      :label="t.FORM.PASSWORD.LABEL"
      :required="required"/>

    <div class="confirmPassword__strengthMeter" :class="{isVisible: strengthBarVisible}">
      <div class="confirmPassword__strengthMeter__text"
      >{{ t.FORM.VALIDATION.PASSWORD_STRENGTH.LEVELS[strengthIndex] }}</div>
      <div class="confirmPassword__strengthMeter__gauge">
        <div class="confirmPassword__strengthMeter__gauge__bar" :class="barClasses"></div>
      </div>
    </div>

    <hd-input
      ref="passwordConfirm"
      v-model="passwordConfirm"
      name="passwordConfirm"
      class="confirmPassword__input"
      type="password"
      :icon="icon"
      :label="t.FORM.PASSWORD.LABEL_CONFIRM"
      :required="required"/>
  </div>
</template>

<script>
import merge from 'lodash/merge';
import { getMessages } from 'hd-blocks/lang';
import { getPasswordStrength } from 'hd-blocks/services/utils';
import HdInput from 'hd-blocks/components/form/HdInput.vue';

export default {
  name: 'hd-confirm-password',
  components: {
    HdInput,
  },
  props: {
    value: {
      type: String,
      default: '',
    },
    name: {
      type: String,
      default: '',
    },
    required: {
      type: Boolean,
      default: true,
    },
    min: {
      type: Number,
      default: 0,
    },
    withStrength: {
      type: Boolean,
      default: true,
    },
    strengthBarStyle: {
      type: Boolean,
      default: false,
    },
    icon: {
      type: String,
      default: '',
    },
    lang: String,
    texts: {
      type: Object,
      default: () => ({}),
    },
  },
  data() {
    return {
      passwordMain: this.value,
      passwordConfirm: '',
      strengthIndex: null,
      strengthBarVisible: false,
    };
  },
  watch: {
    value(value) {
      this.passwordMain = value;
    },
    passwordMain(value) {
      this.$emit('input', value);
      this.$nextTick(this.validate);
    },
    passwordConfirm() {
      this.$nextTick(this.validate);
    },
  },
  computed: {
    t() {
      return merge(getMessages(this.lang), this.texts);
    },
    barClasses() {
      return {
        [`level--${this.strengthIndex}`]: this.strengthBarVisible && this.strengthIndex >= 0,
      };
    },
  },
  methods: {
    validate() {
      const inputsAreMatching = this.checkPasswordMatching();
      const passwordIsLongEnough = this.checkPasswordLength();

      const isValid = inputsAreMatching && passwordIsLongEnough;

      if (this.withStrength && passwordIsLongEnough) {
        this.updatePasswordStrength();
      } else {
        this.strengthBarVisible = false;
      }
      return isValid;
    },
    checkPasswordLength() {
      if (this.passwordMain.length === 0 && this.required) {
        this.$refs.passwordMain.showError(this.t.FORM.VALIDATION.REQUIRED);
        return false;
      }
      if (this.min > this.passwordMain.length) {
        this.$refs.passwordMain.showError(this.t.FORM.VALIDATION.PASSWORD_SHORT);
        return false;
      }
      this.$refs.passwordMain.hideError();
      return true;
    },
    checkPasswordMatching() {
      if (this.passwordConfirm.length === 0) {
        return false;
      }
      if (this.passwordMain !== this.passwordConfirm) {
        this.$refs.passwordConfirm.showError(this.t.FORM.VALIDATION.PASSWORD_MISMATCH);
        return false;
      }
      this.$refs.passwordConfirm.hideError();
      return true;
    },
    updatePasswordStrength() {
      this.strengthIndex = getPasswordStrength(this.passwordMain, 4);
      if (this.strengthBarStyle) {
        this.strengthBarVisible = true;
      } else {
        const COLORS = [
          'orange',
          'goldenRod',
          'limeGreen',
          'green',
        ];
        const { TITLE, LEVELS } = this.t.FORM.VALIDATION.PASSWORD_STRENGTH;
        const msg = `${TITLE}: <span style='color: ${COLORS[this.strengthIndex]}'
        >${LEVELS[this.strengthIndex]}</span>`;
        this.$refs.passwordMain.showHelper(msg);
      }
    },
  },
};
</script>

<style lang="scss" scoped>
@import 'hd-blocks/styles/mixins.scss';

.confirmPassword {
  width: 100%;
  position: relative;
  &__strengthMeter {
    height: 14px;
    width: 100%;
    padding-left: $inline-s;
    position: absolute;
    top: 66px; // input height + margin
    display: flex;
    align-items: center;
    visibility: hidden;
    &.isVisible {
      visibility: visible;
    }
    &__text {
      @include font('text-xxsmall');
      color: $regent-gray;
      width: 65px;
    }
    &__gauge {
      height: 2px;
      flex: 1;
      &__bar {
        background: grey;
        width: 0%;
        height: 100%;
        transition: width .5s, background-image 1s;
        position: relative;
        overflow: hidden;
        &:after {
          content: '';
          height: 100%;
          width: 20px;
          background-image: linear-gradient(to right, rgba(255, 255, 255, 0), rgba(255, 255, 255, .7));
          position: absolute;
          right: -30px;
          top: 0;
          opacity: 0;
        }
        &.level {
          &--0 {
            width: 25%;
            background-image: linear-gradient(to right, #f70, #fa0);
          }
          &--1 {
            width: 50%;
            background-image: linear-gradient(to right, #db0, #bd0);
          }
          &--2 {
            width: 75%;
            background-image: linear-gradient(to right, #9d0, #5d1);
          }
          &--3 {
            &:after {
              animation: fadeInLeft 2s;
            }
            width: 100%;
            background-image: linear-gradient(to right, #1d0, #094);
          }
        }
      }
    }
  }
  @keyframes fadeInLeft {
    0% {
      opacity: 0;
      transform: translateX(-300px);
    }
    20% {
      opacity: 1;
    }
    100% {
      opacity: 1;
      transform: translateX(0px);
    }
  }
}
</style>
