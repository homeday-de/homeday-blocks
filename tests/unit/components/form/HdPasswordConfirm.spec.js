import { mount } from '@vue/test-utils';
import HdPasswordConfirm from '@/components/form/HdPasswordConfirm.vue';
import HdInputPassword from '@/components/form/HdInputPassword.vue';

const STRENGTH_METER_SELECTOR = '.confirmPassword__strengthMeter';
const STRENGTH_BAR_VISIBLE_CLASS = 'isVisible';

const TEST_PASSWORD = 'loremIpsum123';

describe('HdPasswordConfirm', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = mount(HdPasswordConfirm, {
      propsData: { value: TEST_PASSWORD },
      sync: false, // Added to avoid this issue : https://github.com/vuejs/vue-test-utils/issues/532
    });
  });

  it('is rendered as expected', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('has two HdInputPassword inputs', () => {
    expect(wrapper.findAll(HdInputPassword).length).toEqual(2);
  });

  it('shows password strength as text by default', async () => {
    await wrapper.vm.validate();

    expect(wrapper.find(STRENGTH_METER_SELECTOR).text()).toBeTruthy();
  });

  it('does not show the password strength if the prop `withStrength` is set to false', async () => {
    wrapper.setProps({ withStrength: false });

    await wrapper.vm.validate();

    expect(wrapper.find(STRENGTH_METER_SELECTOR).text()).toBeFalsy();
  });

  it('shows password strength as a bar if the prop `strengthBarStyle` is set to true', async () => {
    expect(wrapper.find(STRENGTH_METER_SELECTOR).classes()).not.toContain(STRENGTH_BAR_VISIBLE_CLASS);

    wrapper.setProps({ strengthBarStyle: true });

    await wrapper.vm.validate();

    expect(wrapper.find(STRENGTH_METER_SELECTOR).classes()).toContain(STRENGTH_BAR_VISIBLE_CLASS);
  });
});
