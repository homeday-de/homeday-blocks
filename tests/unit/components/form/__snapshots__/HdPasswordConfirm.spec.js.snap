// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HdPasswordConfirm is rendered as expected 1`] = `
<div class="confirmPassword">
  <div class="password-input confirm-password__input">
    <div class="field field--input field--filled field--hasControl">
      <!----> <input autocomplete="on" type="password" id="passwordMain" name="passwordMain" placeholder="" required="required" data-lpignore="true" class="field__input"> <label for="passwordMain" class="field__label">
        Ihr Passwort
      </label>
      <!----> <span class="field__border"></span></div> <span class="field__visibilityToggle"><!----></span>
  </div>
  <div class="confirmPassword__strengthMeter">
    <div class="confirmPassword__strengthMeter__text"></div>
    <div class="confirmPassword__strengthMeter__gauge">
      <div class="confirmPassword__strengthMeter__gauge__bar"></div>
    </div>
  </div>
  <div class="password-input confirmPassword__input">
    <div class="field field--input">
      <!----> <input autocomplete="on" type="password" id="passwordConfirm" name="passwordConfirm" placeholder="" required="required" data-lpignore="true" class="field__input"> <label for="passwordConfirm" class="field__label">
        Best√§tigen Sie Ihr Passwort
      </label>
      <!----> <span class="field__border"></span></div>
    <!---->
  </div>
</div>
`;
